File: scripts/generate_pdf.js
const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

const MEMOS_DIR = path.resolve(__dirname, '../pages/memos');
const OUTPUT_DIR = path.resolve(__dirname, '../dist_pdf');

// List of the 12 main countries from the index.html homepage
const MAIN_COUNTRIES = [
    'egypt.html',
    'maldives.html',
    'turkey.html',
    'vietnam.html',
    'china.html',
    'mauritius.html',
    'thailand.html',
    'seychelles.html',
    'indonesia.html',
    'sri-lanka.html',
    'tanzania.html',
    'tunisia.html'
];

if (!fs.existsSync(OUTPUT_DIR)) {
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });
}

(async () => {
  console.log('--- Emoji Tours PDF Generator (12 main countries) ---');
  
  let browser;
  try {
    browser = await chromium.launch();
    const page = await browser.newPage();

    // Filter to only include the main countries
    const files = fs.readdirSync(MEMOS_DIR).filter(f => MAIN_COUNTRIES.includes(f.toLowerCase()));

    console.log(`Processing ${files.length} documents...`);

    for (const file of files) {
      // Normalize naming (e.g. sri-lanka.html -> Memo_SriLanka.pdf)
      const baseName = file.replace('.html', '').replace('-', '_');
      const parts = baseName.split('_');
      const prettyName = parts.map(p => p.charAt(0).toUpperCase() + p.slice(1)).join('');
      
      const filePath = `file://${path.join(MEMOS_DIR, file)}`;
      const outputPath = path.join(OUTPUT_DIR, `Memo_${prettyName}.pdf`);

      console.log(`[Processing] ${prettyName}...`);
      
      try {
        await page.goto(filePath, { waitUntil: 'networkidle', timeout: 45000 });
        await page.waitForTimeout(2000);

        await page.pdf({
          path: outputPath,
          format: 'A4',
          printBackground: true,
          margin: { top: '0', bottom: '0', left: '0', right: '0' }
        });
        
        console.log(`  ✓ Generated: Memo_${prettyName}.pdf`);
      } catch (err) {
        console.error(`  ✕ Error in ${file}:`, err.message);
      }
    }

    console.log('Finished.');

  } catch (error) {
    console.error('Critical Error:', error);
  } finally {
    if (browser) await browser.close();
  }
})();

File: scripts/generate_pdf.py
import os
import asyncio
from playwright.async_api import async_playwright

ALLOWED_COUNTRIES = [
    'egypt', 'maldives', 'turkey', 'vietnam', 'china', 
    'mauritius', 'thailand', 'seychelles', 'indonesia', 
    'sri-lanka', 'tanzania', 'tunisia'
]

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
MEMOS_DIR = os.path.join(BASE_DIR, 'pages', 'memos')
OUTPUT_DIR = os.path.join(BASE_DIR, 'dist_pdf')

if not os.path.exists(OUTPUT_DIR):
    os.makedirs(OUTPUT_DIR)

async def generate_pdf(browser, html_file):
    page = await browser.new_page()
    file_path = os.path.join(MEMOS_DIR, html_file)
    normalized_path = file_path.replace(os.sep, '/')
    file_url = f"file:///{normalized_path}"
    
    print(f"--- Начинаю: {html_file} ---")
    try:
        # ИСПОЛЬЗУЕМ wait_until='load', чтобы не ждать бесконечно все картинки/скрипты
        # Это решит проблему с таймаутом на тяжелых страницах (Египет)
        await page.goto(file_url, wait_until="load", timeout=60000)
        
        # Даем 2 секунды на рендеринг шрифтов и базовых стилей
        await asyncio.sleep(2)
        
        output_filename = html_file.replace('.html', '.pdf')
        output_path = os.path.join(OUTPUT_DIR, output_filename)
        
        await page.pdf(path=output_path, format="A4", print_background=True, 
                       margin={"top": "0", "bottom": "0", "left": "0", "right": "0"},
                       display_header_footer=False, scale=0.9)
        print(f"Готово: {output_filename}")
    except Exception as e:
        print(f"Ошибка при работе с {html_file}: {e}")
    finally:
        await page.close()

async def main():
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        memos = [f for f in os.listdir(MEMOS_DIR) if f.endswith('.html')]
        target_memos = [f for f in memos if f.replace('.html', '') in ALLOWED_COUNTRIES]
        
        print(f"Найдено памяток для обработки (FINAL): {len(target_memos)}")
        for memo in target_memos:
            await generate_pdf(browser, memo)
        await browser.close()
        print("\\nВсе PDF успешно сгенерированы!")

if __name__ == "__main__":
    asyncio.run(main())
